Add XboxOne/DualShock layouts for Android

rytis-buz opened the pull request:
Added DualShock4GamepadAndroid and XboxOneGamepadAndroid layouts that derive from DualShockGamepad and XInputController respectively instead of deriving from AndroidGamepad
Fixed Right stick to use AXIS.Z and AXIS.RZ
Fixed triggers to always use Axis.Gas and Axis.Brake
Moved AndroidGamepadWithDpadButtons and AndroidGamepadWithDpadAxes from Json definition to class definition

rytis-buz committed:
Fix Xbox controller recognition

rytis-buz committed:
Add PS4 DualShock layout for Android

rytis-buz committed:
Add XboxOne controller layout deriving from XInput

rytis-buz committed:
clean comment

rytis-buz committed:
Add Dpad variant instead of new state struct

rytis-buz committed:
Add dpad to Dpad variant only and add test

Commit:
FIX: Layout variant separators mixing ',' and ';'.

rytis-buz committed:
Merge branch 'develop' of github.com:Unity-Technologies/InputSystem into mobile/controller-input-fixes

rytis-buz committed:
Enable test which was fixed by adding Dualshock layout for Android

rytis-buz committed:
Merge branch 'develop' of github.com:Unity-Technologies/InputSystem into mobile/controller-input-fixes

rytis-buz committed:
Use DualShockGamepad instead of DualShock4GamepadHID for tests to be able to run them on all platforms also re-enable these tests

rytis-buz committed:
Merge branch 'develop' of github.com:Unity-Technologies/InputSystem into mobile/controller-input-fixes

rytis-buz committed:
Merge branch 'develop' of github.com:Unity-Technologies/InputSystem into mobile/controller-input-fixes

rytis-buz committed:
Fix tests depending on new AndroidXbox layout

rytis-buz committed:
Fix generic controllers with Dpad

rytis-buz committed:
Clamp trigger values and add DualShock tests

rytis-buz commented on Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidGameController.cs:
```diff
@@ -34,18 +42,26 @@ internal unsafe struct AndroidGameControllerState : IInputStateTypeInfo
+        [InputControl(name = "dpad/left", offset = 0, bit = 0, sizeInBits = 32, format = "FLT", parameters = "clamp=3,clampConstant=0,clampMin=-1,clampMax=0,invert", variants = kVariantDPadAxes)]
+        [InputControl(name = "dpad/down", offset = ((uint)AndroidAxis.HatY - (uint)AndroidAxis.HatX) * sizeof(float), bit = 0, sizeInBits = 32, format = "FLT", parameters = "clamp=3,clampConstant=0,clampMin=0,clampMax=1", variants = kVariantDPadAxes)]
+        [InputControl(name = "dpad/up", offset = ((uint)AndroidAxis.HatY - (uint)AndroidAxis.HatX) * sizeof(float), bit = 0, sizeInBits = 32, format = "FLT", parameters = "clamp=3,clampConstant=0,clampMin=-1,clampMax=0,invert", variants = kVariantDPadAxes)]
+        [InputControl(name = "leftTrigger", offset = (uint)AndroidAxis.Brake * sizeof(float) + kAxisOffset, parameters = "clamp=1,clampMin=0,clampMax=1.0", variants = kVariantGamepad)]
```
This clamping for triggers is questionable. There is a test that sets -1 and expects 0 as a result with a comment "Check if normalization works correctly". Clamping seemed to be the way to achieve that

rytis-buz committed:
Split HID tests from generic scheme matching tests to run it on more platforms

rytis-buz committed:
Fix and enable remaining Android/iOS tests

todi1856 commented:
Could you add QA, since this is a sensitive change

todi1856 commented:
Missing entries in CHANGELOG

todi1856 commented on Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidGameController.cs:
```diff
@@ -34,18 +42,26 @@ internal unsafe struct AndroidGameControllerState : IInputStateTypeInfo
         [InputControl(name = "rightStick/y", offset = ((uint)AndroidAxis.Rz - (uint)AndroidAxis.Z) * sizeof(float), variants = kVariantGamepad, parameters = "invert")]
-        [InputControl(name = "rightStick/up", offset = ((uint)AndroidAxis.Rz - (uint)AndroidAxis.Z) * sizeof(float), variants = kVariantGamepad, parameters = "invert,clamp=1,clampMin=-1.0,clampMax=0.0")]
-        [InputControl(name = "rightStick/down", offset = ((uint)AndroidAxis.Rz - (uint)AndroidAxis.Z) * sizeof(float), variants = kVariantGamepad, parameters = "invert=false,clamp=1,clampMin=0,clampMax=1.0")]
+        [InputControl(name = "rightStick/up", offset = ((uint)AndroidAxis.Rz - (uint)AndroidAxis.Z) * sizeof(float), variants = kVariantGamepad,
```
no change here ? can you revert, same for  rightStick

todi1856 commented on Packages/com.unity.inputsystem/InputSystem/Plugins/Android/AndroidGameController.cs:
```diff
@@ -152,13 +168,13 @@ namespace UnityEngine.InputSystem.Android
     ///  Xbox Gamepad (for ex., Microsoft X-Box One pad (Firmware 2015)) mapping (Note mapping: L2/R2/Right Stick)
     ///  Left Stick -> AXIS_X(0) / AXIS_Y(1)
-    ///  Right Stick -> AXIS_RX (12) / AXIS_RY(13)
+    ///  Right Stick -> AXIS_Z (11) / AXIS_RZ(14)
```
So it seems, this comment is no longer relevent, as mapping matches the comment scope above next to /// Most of the gamepads: 

if so, can you delete this scope, and add Xbox Controller in the list above (after Sony Interactive Entertainment Wireless), also could you please add device name and android os version next to the device just as an additional info

todi1856 reviewed (changes requested):
Looks good, tested myself Dualschock on Google Pixel 2 XL with Android 11, works fine.

Please add QA, after QA approves, I'll approve too

todi1856 reviewed (changes requested):
Please add entries to CHANGELOG

rytis-buz committed:
Added comments, some formating changes

rytis-buz committed:
Update changelog

rytis-buz committed:
Formating

renaldasrep commented:
Sorry it's taking a while, we're trying to get a hold of required controllers, should be able to have them at hand by the end of the week

rytis-buz committed:
Update documentation about ps4 limitations on Android

rytis-buz committed:
Add comments about tested devices

renaldasrep reviewed (approved):
Tested on multiple android devices using DualShock 4 and the Xbox Series X controller
No issues found apart DualShock 4 working pretty randomly on Android 8, but it is documented now that it is not officially supported (notes added in this PR)

rytis-buz commented:
@Rene-Damm The PR has been open for 2 months and now have conflicts. Can we move forward with it?

andrew-oc committed:
Merge branch 'develop' into mobile/controller-input-fixes

andrew-oc committed:
Minor changes

Change playstation to PlayStation in SupportedDevices.md to fix nda keyword verification failure

andrew-oc committed:
More tweaks to fix nda verification failure

ekcoh commented:
@rytis-buz @todi1856 @renaldasrep sorry for the late follow-up on this PR but we (@andrew-oc, @ekcoh)  have looked through it now and would be happy to merge it to develop. We did some minor modifications on our end and just want to check that yo uare onboard with this happening now before we merge.

todi1856 commented:
looks good

rytis-buz commented:
Looks good to me
